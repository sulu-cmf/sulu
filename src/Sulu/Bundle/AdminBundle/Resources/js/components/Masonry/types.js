// @flow
export type MasonryItem = {
    id: string | number,
    selected: boolean,
    onClick?: (itemId: string | number) => void,
    onSelectionChange?: (itemId: string | number, checked: boolean) => void,
};
