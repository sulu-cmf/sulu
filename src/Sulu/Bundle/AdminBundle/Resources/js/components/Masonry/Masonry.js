// @flow
import type {ChildrenArray, Node, ElementRef} from 'react';
import React from 'react';
import imagesLoaded from 'imagesloaded';
import MasonryLayout from 'masonry-layout';
import masonryStyles from './masonry.scss';

const MASONRY_OPTIONS = {
    gutter: 20,
    horizontalOrder: true,
    transitionDuration: 250,
};

type Props = {
    children?: ChildrenArray<Node>,
};

export default class Masonry extends React.PureComponent<Props> {
    masonryRef: ElementRef<'div'>;

    masonry: MasonryLayout;

    layoutedChildNodes: Node[];

    componentDidMount() {
        this.initMasonryLayout();
        this.handleImagesLoading();
    }

    componentWillUnmount() {
        this.layoutedChildNodes = [];

        this.destroyMasonry();
    }

    componentDidUpdate() {
        this.handleChildrenUpdates();
        this.handleImagesLoading();
    }

    setMasonryRef = (ref: ElementRef<'div'>) => {
        this.masonryRef = ref;
    };

    getChildNodes() {
        const childNodes = this.masonryRef.children;

        return Array.from(childNodes);
    }

    initMasonryLayout() {
        this.masonry = new MasonryLayout(
            this.masonryRef,
            MASONRY_OPTIONS
        );

        this.layoutedChildNodes = this.getChildNodes();
    }

    destroyMasonry() {
        if (this.masonry) {
            this.masonry.destroy();
            this.masonry = null;
        }
    }

    cloneItems(originalItems: ChildrenArray<*>) {
        const itemStyle = {marginBottom: MASONRY_OPTIONS.gutter};

        return React.Children.map(originalItems, (item) => (
            <li style={itemStyle}>
                {
                    React.cloneElement(
                        item,
                        {
                            key: item.key,
                        }
                    )
                }
            </li>
        ));
    }

    handleChildrenUpdates() {
        const currentChildNodes = this.getChildNodes();
        const knownChildNodes = currentChildNodes.filter((currentChildNode) => {
            return this.layoutedChildNodes.includes(currentChildNode);
        });

        const newChildNodes = currentChildNodes.filter((currentChildNode) => {
            return !knownChildNodes.includes(currentChildNode);
        });

        const removedChildNodes = knownChildNodes.filter((knownChildNode) => {
            return !currentChildNodes.includes(knownChildNode);
        });

        let startIndex = 0;
        const prependedChildNodes = newChildNodes.filter((newChildNode) => {
            const isPrepended = (startIndex === currentChildNodes.indexOf(newChildNode));

            if (isPrepended) {
                startIndex++;
            }

            return isPrepended;
        });

        const appendedChildNodes = newChildNodes.filter((newChildNode) => {
            return !prependedChildNodes.includes(newChildNode);
        });

        if (removedChildNodes.length > 0) {
            this.masonry.remove(removedChildNodes);
        }

        if (appendedChildNodes.length > 0) {
            this.masonry.appended(appendedChildNodes);
        }

        if (prependedChildNodes.length > 0) {
            this.masonry.prepended(prependedChildNodes);
        }

        this.layoutedChildNodes = currentChildNodes;

        if (
            removedChildNodes.length > 0 ||
            appendedChildNodes.length > 0 ||
            prependedChildNodes.length > 0
        ) {
            this.masonry.reloadItems();
        }

        this.masonry.layout();
    }

    handleImagesLoading() {
        imagesLoaded(this.layoutedChildNodes)
            .once('always', () => this.masonry.layout());
    }

    render() {
        const {
            children,
        } = this.props;
        const clonedItems = this.cloneItems(children);

        return (
            <div
                ref={this.setMasonryRef}
                className={masonryStyles.masonry}
            >
                {clonedItems}
            </div>
        );
    }
}
