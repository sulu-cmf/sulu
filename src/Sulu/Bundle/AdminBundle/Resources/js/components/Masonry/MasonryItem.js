// @flow
import type {ElementRef, Node} from 'react';
import React from 'react';
import masonryItemStyles from './masonryItem.scss';

type Props = {
    children: Node,
    itemRef?: ElementRef<'li'>,
};

export default class MasonryItem extends React.PureComponent<Props> {
    render() {
        const {
            itemRef,
            children,
        } = this.props;

        return (
            <li
                ref={itemRef}
                className={masonryItemStyles.masonryItem}>
                {children}
            </li>
        );
    }
}
