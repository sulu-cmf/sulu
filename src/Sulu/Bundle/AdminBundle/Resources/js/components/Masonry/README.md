The `Masonry` component just serves as a container. The children of the Masonry could be any kind of React component. 
The only requirement for the Masonry to work correctly is that every child needs a unique invariable `key`, which
disqualifies the `index` argument inside a `map` callback. The component recognizes appended, prepended and removed 
child elements and adjusts the masonry layout appropriately.
**Note:** Adding elements inbetween of existing children may cause incomprehensible and/or inconsistent behaviour. 

Here a basic example of the Masonry view:

```
initialState = {
    items: [
        { id: 1, color: '#4a86e8', height: 300 },
        { id: 2, color: '#bb5ac4', height: 200 },
        { id: 3, color: '#5ac4b2', height: 150 },
    ]
};

const createItems = () => {
    let id = Date.now();

    return [
        { id: id + 1, color: '#c4825a', height: 200 },
        { id: id + 2, color: '#96c45a', height: 250 },
        { id: id + 3, color: '#c45a72', height: 180 },
    ];
};

const prepend = () => {
    state.items.unshift(...(createItems()));

    setState({
        items: state.items,
    });
};

const append = () => {
    state.items.push(...(createItems()));

    setState({
        items: state.items,
    });
};

const remove = () => {
    setState({
        items: state.items.filter((item, index) => index !== 3),
    });
};

<div>
    <div style={{marginBottom: 30}}>
        <button onClick={() => prepend()}>Prepend</button>        
        <button onClick={() => append()}>Append</button>
        <button onClick={() => remove()}>Remove</button>
    </div>
    <Masonry>
        {
            state.items.map((item) => {
                return (
                    <div
                        key={item.id}
                        style={{
                            width: '260px',
                            height: item.height,
                            borderRadius: 6,
                            backgroundColor: item.color,
                        }}
                    />
                )
            })
        }
    </Masonry>
</div>
```
