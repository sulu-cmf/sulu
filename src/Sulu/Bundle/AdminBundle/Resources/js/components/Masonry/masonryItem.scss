.masonry-item {
    background: #ccc;
    width: 200px;
    padding: 10px;
    margin-bottom: 10px;
    float: left;
}
