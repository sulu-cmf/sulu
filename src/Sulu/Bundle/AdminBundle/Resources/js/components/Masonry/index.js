// @flow
import Masonry from './Masonry';
import MasonryItem from './MasonryItem';

export {
    Masonry,
    MasonryItem,
};
