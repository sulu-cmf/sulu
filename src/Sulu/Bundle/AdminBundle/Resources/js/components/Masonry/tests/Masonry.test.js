/* eslint-disable flowtype/require-valid-file-annotation */
import {render} from 'enzyme';
import React from 'react';
import Masonry from '../Masonry';

test('Render an empty Masonry container', () => {
    expect(render(<Masonry />)).toMatchSnapshot();
});

test('Render a Masonry container with items', () => {
    expect(render(
        <Masonry>
            <div>
                <img src="http://lorempixel.com/300/200" />
            </div>
            <div>
                <img src="http://lorempixel.com/300/200" />
            </div>
            <div>
                <img src="http://lorempixel.com/300/200" />
            </div>
        </Masonry>
    )).toMatchSnapshot();
});
