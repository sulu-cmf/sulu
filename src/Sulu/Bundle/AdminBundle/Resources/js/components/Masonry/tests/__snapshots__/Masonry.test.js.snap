// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a Masonry container with items 1`] = `
<div
  class="masonry"
>
  <li
    style="margin-bottom:20px;"
  >
    <div>
      <img
        src="http://lorempixel.com/300/200"
      />
    </div>
  </li>
  <li
    style="margin-bottom:20px;"
  >
    <div>
      <img
        src="http://lorempixel.com/300/200"
      />
    </div>
  </li>
  <li
    style="margin-bottom:20px;"
  >
    <div>
      <img
        src="http://lorempixel.com/300/200"
      />
    </div>
  </li>
</div>
`;

exports[`Render an empty Masonry container 1`] = `
<div
  class="masonry"
/>
`;
